src/target/riscv/riscv-013.c

```diff
@@ -4394,182 +4394,299 @@ static int write_memory_bus_v1(struct target *target, target_addr_t address,
-			else if (dmi_busy_encountered)
-				LOG_TARGET_DEBUG(target, "Memory write resulted in DMI busy response.");
-			riscv013_clear_abstract_error(target);
-			increase_ac_busy_delay(target);
+static int write_memory_progbuf_inner_fill_progbuf(struct target *target,
+		uint32_t size, bool mprven)
+{
+	if (riscv_save_register(target, GDB_REGNO_S0) != ERROR_OK)
+		return ERROR_FAIL;
+	if (riscv_save_register(target, GDB_REGNO_S1) != ERROR_OK)
+		return ERROR_FAIL;
 
+	struct riscv_program program;
+
+	riscv_program_init(&program, target);
+	if (riscv_program_store_mprv(&program, GDB_REGNO_S1, GDB_REGNO_S0, 0, size,
+				mprven) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (riscv_program_addi(&program, GDB_REGNO_S0, GDB_REGNO_S0, size) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (riscv_program_ebreak(&program) != ERROR_OK)
+		return ERROR_FAIL;
+
+	return riscv_program_write(&program);
+}
+
+static int write_memory_progbuf_inner(struct target *target, target_addr_t address,
+		uint32_t size, uint32_t count, const uint8_t *buffer, bool mprven)
+{
+	if (write_memory_progbuf_inner_fill_progbuf(target, size,
+				mprven) != ERROR_OK)
+		return ERROR_FAIL;
+
+	target_addr_t address_on_target = address;
+	if (write_memory_progbuf_inner_startup(target, &address_on_target, buffer, size) != ERROR_OK)
+		return ERROR_FAIL;
+
+	for (target_addr_t end_address_on_target = address + (target_addr_t)size * count,
+			next_address_on_target = address_on_target;
+			address_on_target != end_address_on_target;
+			address_on_target = next_address_on_target) {
+		const uint8_t * const curr_buff = buffer + (address_on_target - address);
+		if (write_memory_progbuf_inner_try_to_write(target, &next_address_on_target,
+					end_address_on_target, size, curr_buff) != ERROR_OK) {
 			dm_write(target, DM_ABSTRACTAUTO, 0);
-			result = register_read_direct(target, &cur_addr, GDB_REGNO_S0);
-			if (result != ERROR_OK)
-				goto error;
-			setup_needed = true;
-		} else {
-			LOG_TARGET_ERROR(target, "Error when writing memory, abstractcs=0x%08lx", (long)abstractcs);
-			riscv013_clear_abstract_error(target);
-			result = ERROR_FAIL;
-			goto error;
+			return ERROR_FAIL;
+		}
+		if (next_address_on_target == address_on_target) {
+			if (write_memory_progbuf_inner_ensure_forward_progress(target,
+						address_on_target, size, curr_buff) != ERROR_OK) {
+				dm_write(target, DM_ABSTRACTAUTO, 0);
+				return ERROR_FAIL;
+			}
+			next_address_on_target += size;
 		}
+		assert(next_address_on_target - address <= (target_addr_t)size * count);
 	}
 
-error:
-	dm_write(target, DM_ABSTRACTAUTO, 0);
+	return dm_write(target, DM_ABSTRACTAUTO, 0);
+}
+
+static int write_memory_progbuf(struct target *target, target_addr_t address,
+		uint32_t size, uint32_t count, const uint8_t *buffer)
+{
+	if (riscv_xlen(target) < size * 8) {
+		LOG_TARGET_ERROR(target, "XLEN (%u) is too short for %" PRIu64 "-bit memory write.",
+				riscv_xlen(target), (uint64_t)size * 8);
```

Comment:
Detail: I believe it is not necessary to cast the `size` to uint64_t.

```suggestion
	assert(size == 1 || size == 2 || size == 4 || size == 8);
	if (riscv_xlen(target) < size * 8) {
		LOG_TARGET_ERROR(target, "XLEN (%u) is too short for %" PRIu32 "-bit memory write.",
				riscv_xlen(target), size * 8);
```

Reply:
Addressed.